### PR TITLE
TiDB failover

### LINKED ISSUE #59: TiDB automatic failover
When a TiDB peer can't response for a while (can be obtained from tidb instance's `/status` API), operator should:
1. increase the replicas to add a new TiDB peer
2. decrease the replicas when all members are ready again

part of #47

### PR DESCRIPTION
**Pull request by xiaojingchen**
This PR fixes #59 , adds TiDB failover feature

@weekface @tennix @onlymellb @gregwebs  PTAL

### CODE REVIEW

File: pkg/manager/member/tidb_failover.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+			return false
+		}
+	}
+	return true
+}
+
+func (tf *tidbFailover) markThisMemberAsFailure(tc *v1alpha1.TidbCluster, tidbMember v1alpha1.TiDBMember) error {
+	tc.Status.TiDB.FailureMembers[tidbMember.Name] = v1alpha1.TiDBFailureMember{PodName: tidbMember.Name, Replicas: tc.Spec.TiDB.Replicas}
+
+	tc, err := tf.tcControl.UpdateTidbCluster(tc)
```
Comment (weekface) [inline pkg/manager/member/tidb_failover.go]: For TiDB, no need to update TidbCluster immediately.

File: pkg/manager/member/tidb_failover_test.go [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+						Health: true,
+					},
+					"failover-tidb-1": {
+						Name:   "failover-tidb-1",
+						Health: true,
+					},
+				}
+			},
+			expectFn: func(g *GomegaWithT, tc *v1alpha1.TidbCluster) {
+				g.Expect(int(tc.Spec.TiDB.Replicas)).To(Equal(2))
```
Comment (weekface) [inline pkg/manager/member/tidb_failover_test.go]:
```
g.Expect(len(tc.Status.TiDB.FailureMembers)).To(Equal(0))
```

File: cmd/controller-manager/main.go [hunk 3]
```diff
@@ -53,6 +55,8 @@ func init() {
 	flag.IntVar(&workers, "workers", 5, "The number of workers that are allowed to sync concurrently. Larger number = more responsive management, but more CPU (and network) load")
 	flag.BoolVar(&controller.ClusterScoped, "cluster-scoped", true, "Whether tidb-operator should manage kubernetes cluster wide TiDB Clusters")
 	flag.StringVar(&controller.DefaultStorageClassName, "default-storage-class-name", "standard", "Default storage class name")
+	flag.BoolVar(&autoFailover, "auto-failover", false, "Auto failover")
```
Comment (weekface) [inline cmd/controller-manager/main.go]: Add this flag to `charts/tidb-operator/values.yaml`

File: cmd/controller-manager/main.go [hunk 4]
```diff
@@ -53,6 +55,8 @@ func init() {
...
+	flag.DurationVar(&tidbFailoverPeriod, "tidb-failover-period", time.Duration(5*time.Minute), "TiDB failover period default(5m)")
```
Comment (weekface) [inline cmd/controller-manager/main.go]: ditto

File: cmd/controller-manager/main.go (hunk 3 above)
Comment (weekface) [reply cmd/controller-manager/main.go]: And this file `controller-manager-deployment.yaml`

File: cmd/controller-manager/main.go (hunk 4 above)
Comment (weekface) [reply cmd/controller-manager/main.go]: ditto

**Comment by WeiFrank** [comment]
Why this commit not exist within the pingcap repositories now？

**Comment by DanielZhangQD** [comment]
@WeiFrank The commit is in the repo, please check it again.

**Comment by WeiFrank** [comment]
> @WeiFrank The commit is in the repo, please check it again.

Yeah, I have seen  this commit, But I mean this function may be not apply, because i can not see those code in release version.

**Pull request merged**